### ISSUE TITLE
Suggesting new feature: autolabel option for bar plots

### ISSUE THREAD

**Issue by ebauch** [matplotlib/matplotlib#12386]
Whenever I plot some form of bar plot I find myself thinking that it would be nice to have the bars labelled with the actual value counts (or a percentage value). Examples are given in the [barplot gallery](https://github.com/rasbt/matplotlib-gallery#boxplots-and-violinplots): 

![image](https://user-images.githubusercontent.com/3758223/46420899-f6d5ce00-c6fe-11e8-83a1-1b89c6815569.png)

![image](https://user-images.githubusercontent.com/3758223/46421462-1e796600-c700-11e8-9958-217a92666120.png)

This is the code that makes it happen: 

``` 
# label bars
def autolabel(rects):
    for ii,rect in enumerate(rects):
        height = rect.get_height()
        plt.text(rect.get_x()+rect.get_width()/2., 1.02*height, '%s'% (mean_values[ii]),
            ha='center', va='bottom')
autolabel(rects)
```

I think this feature is so crucial that it should be part of barplot via an argument such as ``autolabel = true`` or something. What do people think? Implementing this would be relatively easy I suppose but edge cases (such as error bars) might be harder to deal with. I find myself spending quite some time adjusting the heights and axis limits to make it all fit together - automating this would be a big plus.

**Comment by dstansby** [comment]
I like this idea! It could also just be a standalone function that takes whatever `plt.bar()` outputs, and adds the labels, if we want to avoid adding another keyword argument to `plt.bar()`.

**Comment by jklymak** [comment]
:+1: on an added function versus kwarg, analgous to `clabel` for contours.  Then the function can get its own kwargs specifying label things (rotation, fontsize, offset from bar, clipping, etc).  I'd be :-1: on adding this functionality directly to `bar`.

**Comment by jklymak** [comment]
@ebauch, any interest in a PR submitting a new method to axes: `bar_label`?  You'd need to think through the options you'd like for the actual labels, but they could simply be passed through to the text objects.  Also, please don't use the `plt` interface directly in the code base.  This would go in `lib/matplotlib/axes/_axes.py`

**Comment by ebauch** [comment]
@jklymak happy to do it but I only understood half of the words you said (sorry, I am new to this :)
Whats PR? 
What does it mean to submit a new method to axes? I understand, we want to define a new method of axes called `bar_label` that does something like `def autolabel` above.

What do you mean with not using the plt interface directly in the code?

**Comment by ImportanceOfBeingErnest** [comment]
@ebauch This is mainly explained in the [Contributing code](https://matplotlib.org/devel/contributing.html#contributing-code) and the [Development workflow](https://matplotlib.org/devel/gitwash/development_workflow.html#development-workflow) sections of the [developper's guide](https://matplotlib.org/devel/index.html).   
Creating your first pull request ("PR") will be a little bit of work; if you're happy to take this challenge, we can guide you through in case you experience problems. 

Not using `pyplot` in the code means that inside `lib/matplotlib/axes/_axes.py` you do not have `plt` available. Instead your function would be a method of the `Axes` class and will need to rely on the `Axes`' other methods.

Concerning the actual function there are probably a couple of things to consider to have it working in a generalized manner. 

* It would need to take the labels to show as input, but fall back to the values of the bars in case no labels are provided.
* It would need to work with positive, as well as negative bar heights/widths.
* It should work for bars that don't start at 0, i.e. the height of the patch is different than the position of its farmost end in data coordinates.
* The position of the text should be the end of the bar, plus some padding in units of points. Else you end up with different spacings proportional to the bar height.
* The function should work for vertical `bar`s as well as horizontal `barh`s. (That's a tricky bit, I suppose, because the  `BarContainer` does unfortunately not store any information about the orientation of its patches. So it seems inevitable to let the user specify the orientation, or use a good bunch of heuristics to find that out.)
* kwargs should be passed onto the created `Text` labels.

The signature could look like

    def bar_label(self, rects, labels=None, pad=None, orientation="vertical", **kwargs)

Btw. :+1: for putting a :+1: on your own feature request. GitHub should do that automatically. :laughing:

**Comment by anntzer** [comment]
> The function should work for vertical bars as well as horizontal barhs. (That's a tricky bit, I suppose, because the BarContainer does unfortunately not store any information about the orientation of its patches. So it seems inevitable to let the user specify the orientation, or use a good bunch of heuristics to find that out.)

Actually, you can check the sticky_edges attribute of the rectangles: you'll have y=0 for vertical bars and x=0 for horizontal ones.  (mplcursors uses that info, for example.)  While it is true that one could fiddle with this attribute too, people should just place text themselves if they do that... (IOW, I'd suggest not including the orientation argument, and throw an exception in the case of nonstandard sticky edges.)

**Comment by ebauch** [comment]
Okay - I will give it a shot. Here is some code I found which likely does exactly what we want using axes properties and not plot interface. 

https://stackoverflow.com/questions/33179122/seaborn-countplot-with-frequencies

```
for p in ax.patches:
        ax.annotate('%{:.1f}'.format(p.get_height()), (p.get_x()+0.1, p.get_height()+50))

```

**Comment by liujordan** [comment]
Has this been completed? I'd like to give this a shot if not.

**Comment by jklymak** [comment]
I've not seen a PR for this, and given the comment above is several months old, I'd say we would welcome a PR from anyone at this point...

**Comment by hovikgas** [comment]
Seriously. I'm going crazy running the same loop over and over again to add data/value labels for my bar charts... I've never contributed to any open source projects, but it looks like no one else is doing it, so I guess here goes nothing...

**Comment by mwaskom** [comment]
This is closed by #15602, right?
